cfn-resources/autogen/main.go

```diff
@@ -0,0 +1,402 @@
+
+}
+
+func addResource(path, resourceName, schemaName string) {
+	f := NewFile("resource")
+	funcNames := []string{CreateMethod, ReadMethod, UpdateMethod, DeleteMethod, ListMethod}
+
+	//constants := "github.com/mongodb/mongodbatlas-cloudformation-resources/util/constants"
+	//validations := fmt.Sprintf("github.com/mongodb/mongodbatlas-cloudformation-resources/%s/cmd/validations", resourceName)
+
+	//Imports
+	f.ImportName(handler, "handler")
+	f.ImportName(cloudformation, "cloudformation")
+	f.ImportName("errors", "errors")
+	f.ImportName(util, "util")
+	f.ImportName(validator, "validator")
+	f.ImportName(progressEvent, "progress_events")
+	f.ImportAlias(log, "log")
+
+	//Required Fields mapping
+	reqFields, err := readRequiredParameters(schemaName)
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+
+	if reqFields != nil {
+		f, err = addValidator(f, *reqFields)
+		if err != nil {
+			fmt.Println(err)
+			return
+		}
+	}
+
+	//Add setup method
+	f.Func().Id(setupMethod).Params().
+		Block(
+			Qual(util, "SetupLogger").Call(Lit("mongodb-atlas-" + schemaName)),
+		)
+	f.Line()
+
+	//Add handlers
+	for _, fn := range funcNames {
+		inputParameters, modelObj := generateInputParams(fn, reqFields)
+		if len(inputParameters) > 0 {
+			inputParameters = "Considerable params from currentModel: \n" + inputParameters
+		}
+
+		atlasFunc := " res , resModel, err := client." + capitalize(schemaName) + "." + fn + "(context.Background(),&mongodbatlas." + capitalize(schemaName) + "{\n" + modelObj + "})"
+
+		f.Func().Id(fn).Params(
+			//Input Parameters
+			Id("req").Qual(handler, "Request"),
+			Id("prevModel").Id("*Model"),
+			Id("currentModel").Id("*Model")).
+			//Return types
+			Id("(").Qual(handler, "ProgressEvent").Id(",").Id("error").Id(")").
+
+			// Function starts
+			Block(
+				//log setup()
+				Id("setup").Params(),
+				Line(),
+				//Debug log
+				Qual(log, "Debugf").Params(Lit(fn+"() currentModel:%+v"), Id("currentModel")),
+				Line(),
+
+				//Validator
+				Comment("Validation"),
+				Id("modelValidation").Id(":=").Id(validationMethod).Call(Id(fn+"RequiredFields"), Id("currentModel")),
+				If(Id("modelValidation").Op("!=").Id("nil")).Block(
+					Return(
+						Id("*modelValidation"),
+						Id("nil"),
+					),
+				),
+				Line(),
+
+				//Atlas Client
+				Comment("Create atlas client"),
+				List(Id("client"), Err()).Op(":=").Qual(util, "CreateMongoDBClient").
+					Call(Id("*currentModel.ApiKeys.PublicKey"),
+						Id("*currentModel.ApiKeys.PrivateKey")),
+
+				If(Id("err").Op("!=").Id("nil")).Block(
+					Qual(log, "Debugf").
+						Call(Lit(fn+" - error: %+v"), Id("err")),
+
+					Return(
+						Qual(handler, "ProgressEvent").
+							Values(Dict{
+								Id("OperationStatus"):  Qual(handler, "Failed"),
+								Id("Message"):          Id("err.Error()"),
+								Id("HandlerErrorCode"): Qual(cloudformation, "HandlerErrorCodeInvalidRequest"),
+							}),
+						Id("nil"),
+					),
```

Comment:
with this we can use the other progressEvent fail function
```suggestion
					Return(
						Qual(progressEvent, "GetFailedEventByCode").
							Params(Id("err.Error()"), Qual(cloudformation, "HandlerErrorCodeGeneralServiceException")),
						Id("nil"),
					),
```